File: cpp/puzzle.cpp
/*----------------------------------------------------------------------------
				   Puzzle!
				  ---------
______________________________________________________________________________*/
//Coded by: Rohan Siddhu
//Date: 28/04/2019

/* Files Included */

#include <iostream>
#include <conio.h>
#include <stdlib.h>
#include <windows.h>


/* Global variable declaration */

int key;
short i, j;


/* Function prototype declaration */

short menu();
void playscreen();
void gameover();
void table();
int getkey();
short ascending(short q[][4]);


/* inline function declaration */

 /* This function sets the console cursor position at the coordinates (col, row) */
inline void gotoxy(short col, short row)
{
 HANDLE hStdout=GetStdHandle(STD_OUTPUT_HANDLE); 
 COORD position={col, row};

 SetConsoleCursorPosition(hStdout, position);
}

 /* This function draw a horizontal line of length len. */
inline void hline(short len)
{
 for(j=0; j<len; j++)
 std::cout<<char(196);
}

 /* This function draws a virtical line from coordinates (col, row) of length len. */
inline void vline(short len, short col, short row)
{
 for(j=0; j<len; j++)
 {
  gotoxy(col, row+j);
  std::cout<<char(179);
 }
}

 /* This functio draw a rectangle between the coordinates (a, b) and (c, d) */
inline void rectangle(short a, short b, short c, short d)
{
 gotoxy(a, b);
 std::cout<<char(218);
 for(i=0; i<(c-a-1); i++)
  std::cout<<char(196); 
 std::cout<<char(191);
 for(i=1; i<=(d-b-1); i++)
 {
  gotoxy(c, b+i);
  std::cout<<char(179);
 }
 gotoxy(a, b+1);
 for(i=1; i<=(d-b-1); i++)
 {
  gotoxy(a, b+i);
  std::cout<<char(179);
 }
 gotoxy(a, d);
 std::cout<<char(192);
 for(i=0; i<(c-a-1); i++)
  std::cout<<char(196);
  std::cout<<char(217);
}


/* main functon starts */

int main(int argc, char * argv[])
{
 system("cls");                     // Clearing console window.

 while(1)
 {
  menu();
  system("cls");

  playscreen();
 }

 return 0;
}

/* main function ends */


/* Main menu of the game */

short menu()
{
 short h=0;
 system("cls");

 gotoxy(38, 11);
 std::cout<<"Play";
 rectangle(37, 10, 42, 12);

 gotoxy(38, 14);
 std::cout<<"Exit";


 while(1)
 {
  key=getkey();

  if(key==80 && h==0)                 //Select the box over Exit.
  {
   system("cls");
   gotoxy(38, 11);
   std::cout<<"Play";
   gotoxy(38, 14);
   std::cout<<"Exit";
   rectangle(37, 13, 42, 15);
   h++;
  }

  if(key==13 && h==1)                 //Terminates the execution of the program.
   exit(0);

  if(key==72 && h==1)                 //Select the box over Play.
  {
   system("cls");
   gotoxy(38, 11);
   std::cout<<"Play";
   gotoxy(38, 14);
   std::cout<<"Exit";
   rectangle(37, 10, 42, 12);
   h--;
  }

  if(key==13 && h==0)
   return 0;
 }
}


/* Screen of Game Play */

void playscreen()
{
 short arr[4][4]={
		  {1,4,15,7},
		  {8,10,2,11},
		  {14,3,6,13},
		  {12,9,5,16}
		 };                    //Matrix of the puzzle declairation.
 short x,y,move=0,change;
 short a=3,b=3;

 table();                             //Draw table.

 while(1)
 {
  for(i=0,y=8; i<4; i++,y+=6)              //Printing the matrix.
  {
   for(j=0,x=24; j<4; j++,x+=10)
   {
    gotoxy(x,y);

    if(arr[i][j]==16)                  //Print blank space at the place of 16.
     std::cout<<"  ";
    else
    {
     std::cout.width(2);
     std::cout<<arr[i][j];
    }
   }
  }
  gotoxy(50, 2);
  std::cout<<"Moves:";
  gotoxy(58,2);
  std::cout<<move;

  if(ascending(arr)==0)          //Terminating when the matrix is arranged.
   break;

  key=getkey();                  //Taking controls input.

  if(key==27)                    //Contro is passed to homescreen while ESC is pressed.
  {
   menu();
   system("cls");
   table();
  }
  else if(key==72 && a!=3)         //Up arrow key.
  {
   change=arr[a][b];
   arr[a][b]=arr[a+1][b];
   arr[a+1][b]=change;
   a++;
   move++;
  }
  else if(key==80 && a!=0) //Down arrow key.
  {
   change=arr[a][b];
   arr[a][b]=arr[a-1][b];
   arr[a-1][b]=change;
   a--;
   move++;
  }
  else if(key==75 && b!=3) //Left arrow key.
  {
   change=arr[a][b];
   arr[a][b]=arr[a][b+1];
   arr[a][b+1]=change;
   b++;
   move++;
  }
  else if(key==77 && b!=0) //Right arrow key.
  {
   change=arr[a][b];
   arr[a][b]=arr[a][b-1];
   arr[a][b-1]=change;
   b--;
   move++;
  }
 }
 gameover();
}


/* This screen appears when Game is over */

void gameover()
{
 system("cls");
 gotoxy(35, 12);
 std::cout<<"Congretulations";
 getch();
}


/* This function drow a table of 4X4 on the screen. */

void table()
{
 for(i=19; i<=59; i+=10)
 {
  vline(24, i, 5);
 }
 gotoxy(19, 5);                  //Horizontal Line 1
 std::cout<<char(218);
 hline(9);
 for(i=0; i<3; i++)
 {
  std::cout<<char(194);
  hline(9);
 }
 std::cout<<char(191);

 gotoxy(19, 11);                //Horizontal Line 2
 std::cout<<char(195);
 hline(9);
 for(i=0; i<3; i++)
 {
  std::cout<<char(197);
  hline(9);
 }
 std::cout<<char(180);

 gotoxy(19, 17);               //Horizontal Line 3
 std::cout<<char(195);
 hline(9);
 for(i=0; i<3; i++)
 {
  std::cout<<char(197);
  hline(9);
 }
 std::cout<<char(180);

 gotoxy(19, 23);              //Horizontal Line 4
 std::cout<<char(195);
 hline(9);
 for(i=0; i<3; i++)
 {
  std::cout<<char(197);
  hline(9);
 }
 std::cout<<char(180);

 gotoxy(19, 29);             //Horizontal Line 5
 std::cout<<char(192);
 hline(9);
 for(i=0; i<3; i++)
 {
  std::cout<<char(193);
  hline(9);
 }
 std::cout<<char(217);
}


/* This function return the scan code of key pressed. */

int getkey()
{
 int ch;
 ch=getch();
 if(ch==0)
 {
  ch=getch();
  return ch;
 }
 return ch;
}

/* This function return 1 when array is not in ascending order and return 0
   when array is in ascending order. */

short ascending(short q[][4])
{
 short a;
 for(a=1,i=0; i<4; i++)
  for(j=0; j<4; j++,a++)
   if(q[i][j]!=a)
    return 1;

 return 0;
}

File: python/puzzle.py
'''--------------------------------------------------------------------------------------------------
                                            PUZZLE!
                                          -----------
    _____________________________________________________________________________________________'''

# File: main.py
# Coded By: Rohan Siddhu
# Date: 27/11/2019

''' Files Included '''

import msvcrt       # Microsoft Visual C/C++ Runtime Library
import colorama; colorama.init()     # To use ANSI escape sequence.
import os


''' Function's Declaration '''

clear=lambda: os.system("cls")      # Clear Console Window.

def gotoxy(row, col):
    ''' Move cursor to coordinates x,y '''
    print("\033[%d;%dH"%(row, col), end='')      # Using ANSI escape sequence to position cursor.

def hline(length):
    ''' This function draws horizontal line of given length. '''
    for i in range(length):
        print(chr(45), end='')

def vline(length, row, col):
    ''' This function draws vertical line of given length. '''
    for i in range(length):
        gotoxy(row+i, col)
        print(chr(124), end='')

def rectangle(a, b, c, d):
    ''' This function draw a rectangle between the coordinates (a, b) and (c,d) '''
    gotoxy(a, b)
    print(chr(43), end='')
    for i in range(d-b-1):
        print(chr(45), end='')
    print(chr(43), end='')

    for i in range(1, c-a):
        gotoxy(a+i, b)
        print(chr(124), end='')

    for i in range(1, c-a):
        gotoxy(a+i, d)
        print(chr(124), end='')

    gotoxy(c, b)
    print(chr(43), end='')
    for i in range(d-b-1):
        print(chr(45), end='')
    print(chr(43), end='')

def table():
    ''' Draw a table of 4X4 on console window. '''
    for i in range(19, 60, 10):
        vline(24, 5, i)

    gotoxy(5, 19)       # hline 1
    for i in range(4):
        print(chr(43), end='')
        hline(9)
    print(chr(43))

    gotoxy(11, 19)      # hline 2
    for i in range(4):
        print(chr(43), end='')
        hline(9)
    print(chr(43), end='')

    gotoxy(17, 19)      # hline 3
    for i in range(4):
        print(chr(43), end='')
        hline(9)
    print(chr(43), end='')

    gotoxy(23, 19)      # hline 4
    for i in range(4):
        print(chr(43), end='')
        hline(9)
    print(chr(43), end='')

    gotoxy(29, 19)      # hline 5
    for i in range(4):
        print(chr(43), end='')
        hline(9)
    print(chr(43), end='')

def ascending(arr):
    ''' This function return 1 when the given matrix is not in ascending
        order and return 0 when the matrix is in ascending order.'''
    a=1
    for i in range(4):
        for j in range(4):
            if arr[i][j]!=a:
                return 1
            a+=1
    return 0

def gameover():
    ''' This function is called when game is over. '''
    clear()
    gotoxy(12, 35)
    print("Congretulations! You Win!", end='')
    input()

def menu():
    ''' Main menu of the game. '''
    select=0
    clear()

    gotoxy(11, 38)
    print("Play", end='')
    rectangle(10, 37, 12, 42)

    gotoxy(14, 38)
    print("Exit", end='')

    while(True):
        key=msvcrt.getch()
        if key==b'\x00' or key==b'\xe0':
            key=msvcrt.getch()

        if key==b'P' and select==0:     # Select Exit option.
            clear()
            gotoxy(11, 38)
            print("Play", end='')
            gotoxy(14, 38)
            print("Exit", end='')
            rectangle(13, 37, 15, 42)
            select+=1

        if key==b'\r' and select==1:    #Terminate the execution.
            exit(0)

        if key==b'H' and select==1:     # Select Play option
            clear()
            gotoxy(11, 38)
            print("Play", end='')
            gotoxy(14, 38)
            print("Exit", end='')
            rectangle(10, 37, 12, 42)
            select-=1

        if key==b'\r' and select==0:
            return 0

def play():
    ''' This function control the play screen. '''
    clear()     # This function clear console window.
    move=0; a,b=3,3
    arr=[[1, 4, 15, 7],
         [8, 10, 2, 11],
         [14, 3, 6, 13],
         [12, 9, 5, 16]]    # This is the matrix to be solved.

    table()
    
    while(True):
        ''' Printing matrix. '''
        row=8
        for i in range(4):
            col=24
            for j in range(4):
                gotoxy(row, col)
                col+=10
                if arr[i][j]==16:
                    print("  ", end='')
                else:
                    print("%2d"%(arr[i][j]), end='')
            row+=6
        gotoxy(2, 50)
        print("Moves: ", end='')
        gotoxy(2, 58)
        print(move, end='')

        if ascending(arr)==0:
            break  # Terminate the loop when matrix is arranged.

        key=msvcrt.getch()
        if key==b'\xe0' or key==b'\x00':
            key=msvcrt.getch()

        if key==b'\x1b':    # Control is passed to menu when ESC is pressed. 
            menu()
            clear()
            table()
        elif key==b'H' and a<3:      # Up arrow key.
            arr[a][b],arr[a+1][b]=arr[a+1][b],arr[a][b]
            a+=1; move+=1
        elif key==b'P' and a>0:        # Down arrow key.
            arr[a][b],arr[a-1][b]=arr[a-1][b],arr[a][b]
            a-=1; move+=1
        elif key==b'K' and b<3:        # Left arrow key.
            arr[a][b],arr[a][b+1]=arr[a][b+1],arr[a][b]
            b+=1; move+=1
        elif key==b'M' and b>0:        # Right arrow key.
            arr[a][b],arr[a][b-1]=arr[a][b-1],arr[a][b]
            b-=1; move+=1
    gameover()

def main():
    ''' This is the main function of the program. '''
    clear()

    while(True):
        menu()
        clear()
        play()

if __name__=="__main__":
    main()
